### PR TITLE
[types]: allow `data` field in jsonrpc error obj.

### PR DESCRIPTION
**Pull request by niklasad1**
Follow up on #285 

Basically when a error response with a `data` field is included it was regarded as bad response which this fixes

For example I tested this with `subxt` and the following was regarded as bad response before this PR:

```
[2021-04-24T10:44:15Z DEBUG jsonrpsee_ws_client::client] [backend]: recv unparseable message: Ok(Object({"error": Object({"code": Number(1014), "data": String("The transaction has too low priority to replace another transaction already in the pool."), "message": String("Priority is too low: (12497117750671931 vs 12497117750671931)")}), "id": Number(12), "jsonrpc": String("2.0")}))

```

//cc @tomusdrw I think this fixes your `InvalidRequestId` messages.

### CODE REVIEW

File: types/src/v2/error.rs [hunk 1]
```diff
@@ -160,23 +163,33 @@ impl<'de> Visitor<'de> for ErrorCodeVisitor {
...
+
+		loop {
+			match access.next_entry::<&str, i32>() {
+				Ok(Some((key, val))) if key == ERROR_CODE_KEY && res.is_none() => {
+					res = Some(Ok(val.into()));
+				}
+				Ok(Some((key, _))) if key == ERROR_CODE_KEY => {
+					res = Some(Err(DeserializeError::duplicate_field(ERROR_CODE_KEY)));
+				}
+				Ok(None) => break,
```
Comment (niklasad1) [inline types/src/v2/error.rs:183]: deny unknown fields?!

File: http-server/src/server.rs [hunk 2]
```diff
@@ -186,15 +189,20 @@ impl Server {
...
 							Err(_e) => {
-								let (id, err) = match serde_json::from_slice::<JsonRpcInvalidRequest>(&body) {
-									Ok(req) => (req.id, ErrorCode::InvalidRequest),
-									Err(_) => (None, ErrorCode::ParseError),
+								let (id, code) = match serde_json::from_slice::<JsonRpcInvalidRequest>(&body) {
+									Ok(req) => (req.id, JsonRpcErrorCode::InvalidRequest),
+									Err(_) => (None, JsonRpcErrorCode::ParseError),
 								};
-								send_error(id, &tx, err);
+								send_error(id, &tx, JsonRpcErrorObject { code, message: code.message(), data: None });
```
Comment (dvdplm) [inline http-server/src/server.rs:205]: Maybe we can have a `JsonRpcErrorCode::into_error_obj()` for this? `send_error(id, &tx, code.into_error_obj());`?

File: types/src/v2/error.rs [hunk 3]
```diff
@@ -21,18 +21,50 @@ pub struct JsonRpcError<'a> {
...
 impl fmt::Display for JsonRpcErrorAlloc {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-		write!(f, "{:?}: {}: {:?}", self.jsonrpc, self.error, self.id)
+		write!(f, "{:?}: {:?}: {:?}", self.jsonrpc, self.error, self.id)
 	}
 }
 
+/// JSON-RPC error object.
+#[derive(Debug, PartialEq, Clone, Deserialize)]
+#[serde(deny_unknown_fields)]
```
Comment (dvdplm) [inline types/src/v2/error.rs:38]: Do we really need this?

File: types/src/v2/error.rs [hunk 4]
```diff
@@ -21,18 +21,50 @@ pub struct JsonRpcError<'a> {
...
+	/// Message
+	pub message: String,
+	/// Optional data
+	#[serde(skip_serializing_if = "Option::is_none")]
+	pub data: Option<JsonValue>,
+}
+
+impl From<JsonRpcErrorCode> for JsonRpcErrorObjectAlloc {
+	fn from(code: JsonRpcErrorCode) -> Self {
+		Self { message: code.message().to_owned(), code, data: None }
```
Comment (dvdplm) [inline types/src/v2/error.rs:51]:
```suggestion
		Self { code, message: code.message().to_owned(), data: None }
```
…to keep clippy happy?

File: types/src/v2/error.rs [hunk 5]
```diff
@@ -21,18 +21,50 @@ pub struct JsonRpcError<'a> {
...
+}
+
+/// JSON-RPC error object with no extra allocations.
+#[derive(Debug, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct JsonRpcErrorObject<'a> {
+	/// Code
+	pub code: JsonRpcErrorCode,
+	/// Message
+	pub message: &'a str,
```
Comment (dvdplm) [inline types/src/v2/error.rs:62]: Could this be a `Cow` you think?

**Review by dvdplm — APPROVED** [review]
lgtm modulo nits/suggestions

File: types/src/v2/error.rs (hunk 3 above)
Comment (niklasad1) [reply types/src/v2/error.rs:38]: The reason is the huge error type which requires ownership and we return it the user.

The actual body is read and dropped there maybe we could have like `to_owned` or something for  `JsonRpcErrorObject`, it's quite annoying of to have two different types but perhaps we could tackle that in another PR

File: types/src/v2/error.rs (hunk 5 above)
Comment (niklasad1) [reply types/src/v2/error.rs:61]: the intention is that this type should only be used for serialize the object and `Cow` should not be needed then unless we merge to two different types somehow ^^

File: types/src/v2/error.rs [hunk 6]
```diff
@@ -121,7 +126,7 @@ impl JsonRpcErrorCode {
 	}
 
 	/// Returns the message for the given error code.
-	pub const fn message(&self) -> &str {
+	pub const fn message(&self) -> &'static str {
```
Comment (dvdplm) [inline types/src/v2/error.rs:129]: I guess a `&str` returned from a `const fn` is `'static` by definition, but I don't mind being explicit. 👍

Comment (niklasad1) [reply types/src/v2/error.rs:129]: It didn't work in the `From impl` so I had put `'static` there.

```
impl<'a> From<JsonRpcErrorCode> for JsonRpcErrorObject<'a> {
	fn from(code: JsonRpcErrorCode) -> Self {
		Self { code, message: code.message(), data: None }
	}
}
```

Not sure if there's a better work around for that

**Pull request merged**